kubeflow/trainer/rhai/transformers.py

```diff
@@ -0,0 +1,407 @@
+    progress callback, and Trainer monkey-patching - all without requiring
+    any SDK imports at runtime.
+
+    Args:
+        metrics_port: Port for HTTP metrics server.
+        custom_metrics: Dict mapping log keys to metric names.
+
+    Returns:
+        Python code as string with {{user_func_import_and_call}} placeholder.
+    """
+    import json
+
+    custom_metrics_json = json.dumps(custom_metrics)
+
+    return f'''
+# =============================================================================
+# KUBEFLOW SDK - SELF-CONTAINED INSTRUMENTATION WRAPPER
+# Generated by kubeflow.trainer.rhai.transformers
+# No SDK imports needed - all code is self-contained
+# =============================================================================
+
+import http.server
+import json
+import socketserver
+import threading
+import time
+from datetime import datetime
+from typing import Optional
+
+# =============================================================================
+# HTTP METRICS SERVER
+# =============================================================================
+
+class ProgressMetricsHandler(http.server.BaseHTTPRequestHandler):
+    """HTTP handler that serves training progress metrics as JSON."""
+
+    metrics = {{
+        "status": "initializing",
+        "status_message": "Training initialization in progress",
+        "status_details": {{"last_event": "init", "last_event_time": None, "warnings": []}},
+        "progress": {{"step_current": 0, "step_total": 0, "percent": 0.0, "epoch": 0}},
+        "time": {{
+            "started_sec": None,
+            "started_at": None,
+            "elapsed_sec": 0,
+            "elapsed": "0s",
+            "remaining_sec": None,
+            "remaining": None,
+            "updated_sec": None,
+            "updated_at": None,
+        }},
+        "metrics": {{
+            "loss": None,
+            "learning_rate": None,
+            "throughput_samples_sec": None,
+        }},
+    }}
+    metrics_lock = threading.Lock()
+
+    def log_message(self, format, *args):
+        """Suppress default HTTP server logging."""
+        pass
+
+    def do_GET(self):
+        """Respond to GET requests with health check or metrics."""
+        if self.path == "/health":
+            self.send_response(200)
+            self.send_header("Content-Type", "text/plain")
+            self.end_headers()
+            self.wfile.write(b"OK")
+        elif self.path in ("/", "/metrics"):
+            self.send_response(200)
+            self.send_header("Content-Type", "application/json")
+            self.end_headers()
+            metrics_json = ProgressMetricsHandler.get_metrics_json()
+            self.wfile.write(metrics_json.encode("utf-8"))
+        else:
+            self.send_response(404)
+            self.send_header("Content-Type", "text/plain")
+            self.end_headers()
+            self.wfile.write(b"Not Found")
```

Comment:
suggestion: we don't need to support healthz, so I think we can just ignore the path completely

```suggestion
       self.send_response(200)
       self.send_header("Content-Type", "application/json")
       self.end_headers()
       metrics_json = ...
       self.wfile.write(metrics_json.encode("utf-8"))
```

Reply:
IMO the explicit checks will be needed for debugging purpose incase any issue comes in
I'd be happy to discuss if I've misunderstood you!